### PR TITLE
Feat: Improve pre and post hook impl

### PR DESCRIPTION
**Pull request by z3z1ma**
## Context

From slack

> Something occurred to me on post hooks.
I love the fact you can have pre hooks by simply having valid SQL statements above the SELECT statement. What is to say you can't have post-hooks be valid SQL statements below the SELECT?
I know, you may immediately think oh then we need some token or way to parse which of the array of statements is the actual model contents; but that is not so.
That's because you can simply enforce there is only 1 SELECT statement per model. Running a SELECT statement in a hook is completely useless. Therefore all non-select statements can be ran with the model and their position above or below dictates pre or post.
What triggered this thought was I have a MERGE statement used to update an activity_next_occurrence column containing the surrogate key of the next occurrence of the same event performed by the same actor. Which essentially reduces the compute required for a self join down the line. I figured it'd be pretty ugly to stuff that in the MODEL's post_hook as some janky string

## Details

Keeping the surface area as light as possible, this should work to support hooks more naturally through the existing interfaces where they are ran, `pre` and `post` respectively. Because we now parse pre/post and put it in the expected model metadata -- we no longer need the other code path that just looped through expressions excluding macro defs and ran them. That is basically a pre hook but not going through the prehook code pathway.

### CODE REVIEW

File: sqlmesh/core/model/definition.py [hunk 1]
```diff
@@ -992,6 +994,32 @@ def load_model(
...
+
+    Raises:
+        ConfigError: If the model definition contains more than one SELECT query.
+    """
+    query_positions = [(e, ix) for ix, e in enumerate(expressions) if isinstance(e, exp.Select)]
+
+    if not query_positions:
+        return None, [], []
+    elif len(query_positions) > 1:
+        raise_config_error("Only one SELECT query is allowed per model")
```
Comment (tobymao) [inline sqlmesh/core/model/definition.py:1017]: should specify which model this is for easier debugging

**Comment by tobymao** [comment]
this looks great, can we add a test case?

**Comment by z3z1ma** [comment]
Yeah gotta fix something first, then will add a test or two

File: sqlmesh/core/model/definition.py [hunk 2]
```diff
@@ -919,8 +919,19 @@ def load_model(
...
+    if any(isinstance(e, (exp.Subqueryable, d.Jinja)) for e in expressions):
+        try:
+            # Extract the query and any pre/post statements
+            query, pre, post = extract_sql_model_select(expressions[1:])
+        except Exception as e:
+            raise_config_error(f"Invalid model definition: {e}", path)
+        statements = pre + post
+    else:
+        # Assume this is a different model type (seed, python, etc.)
+        query = expressions[-1] if len(expressions) > 1 else None
```
Comment (tobymao) [inline sqlmesh/core/model/definition.py:932]: seed and python models don't have queries, why did you need this block?

File: sqlmesh/core/model/definition.py [hunk 3]
```diff
@@ -956,6 +967,25 @@ def load_model(
             **meta_fields,
         )
     elif query is not None:
+        # Inject preceding and following statements into the model's meta fields.
+        meta_pre: HookCall = meta_fields.setdefault("pre", exp.Tuple())
+        if not pre:
+            pass  # Nothing parsed to inject
+        elif hasattr(meta_pre, "expressions"):
+            meta_pre.expressions.extend([s for s in pre if not isinstance(s, (d.MacroDef))])
```
Comment (tobymao) [inline sqlmesh/core/model/definition.py:975]: this section is a bit sus, how was this handeled before and why do you need it now

File: sqlmesh/core/snapshot/evaluator.py [hunk 4]
```diff
@@ -136,9 +136,6 @@ def apply(query_or_df: QueryOrDF, index: int = 0) -> None:
                 else:
                     raise SQLMeshError(f"Unexpected SnapshotKind: {snapshot.model.kind}")
 
-        for sql_statement in model.sql_statements:
```
Comment (tobymao) [inline sqlmesh/core/snapshot/evaluator.py:139]: let's delete sql_statements in model definition if it is not used

File: sqlmesh/core/model/definition.py (hunk 3 above)
Comment (tobymao) [reply sqlmesh/core/model/definition.py:975]: i feel like this could probably be cleaned up

File: sqlmesh/core/model/definition.py (hunk 2 above)
Comment (z3z1ma) [reply sqlmesh/core/model/definition.py:932]: I fallback to the codepath that already existed because tests were getting messed up and it was really hard to figure out. The worst offender was the `test_no_query` or something.

File: sqlmesh/core/model/definition.py (hunk 3 above)
Comment (z3z1ma) [reply sqlmesh/core/model/definition.py:975]: I think the issue is that macros are a union type and are already not solidly defined in a single interface

Comment (z3z1ma) [reply sqlmesh/core/model/definition.py:975]: So the pre/post could already be just an `exp.Anonymous` instead of a `exp.Tuple`/`exp.Array` and I need to wrap it basically in order for the downstream parsing to pick up that it needs to iterate through expressions. At least that was what I observed and what made tests happy

Comment (z3z1ma) [reply sqlmesh/core/model/definition.py:975]: Needs a real interface, probably. But maybe a way to get by for now thats cleaner

File: sqlmesh/core/snapshot/evaluator.py (hunk 4 above)
Comment (z3z1ma) [reply sqlmesh/core/snapshot/evaluator.py:139]: Ok I'll take a look, will do 👍

File: sqlmesh/core/model/definition.py (hunk 3 above)
Comment (z3z1ma) [reply sqlmesh/core/model/definition.py:975]: Let me take a second look rq, I think I panicked when I saw the union type. 😆

Comment (z3z1ma) [reply sqlmesh/core/model/definition.py]: Maybe take a second look now, I adjusted the code to what I think is a more ideal state although it might make a few tests angry.

**Comment by tobymao** [comment]
looks better, let's get this in
